programs/keeper-client/KeeperClient.cpp

```diff
@@ -0,0 +1,342 @@
+         }},
+
+        {"ls", 0, [](KeeperClient * client, const std::vector<String> & /* args */)
+         {
+             auto children = client->zookeeper->getChildren(client->cwd);
+             for (auto & child : children)
+                 std::cout << child << " ";
+             std::cout << "\n";
+         }},
+
+        {"ls", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             auto children = client->zookeeper->getChildren(client->getAbsolutePath(args[1]));
+             for (auto & child : children)
+                 std::cout << child << " ";
+             std::cout << "\n";
+         }},
+
+        {"cd", 0, [](KeeperClient * /* client */, const std::vector<String> & /* args */)
+         {
+         }},
+
+        {"cd", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             auto new_path = client->getAbsolutePath(args[1]);
+             if (!client->zookeeper->exists(new_path))
+                 std::cerr << "Path " << new_path << " does not exists\n";
+             else
+                client->cwd = new_path;
+         }},
+
+        {"rm", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             client->zookeeper->remove(client->getAbsolutePath(args[1]));
+         }},
+
+        {"rmr", 1, [](KeeperClient * client, const std::vector<String> & args)
+         {
+             String path = client->getAbsolutePath(args[1]);
+             client->askConfirmation("You are going to recursively delete path " + path,
+                                     [client, path]{ client->zookeeper->removeRecursive(path); });
+         }},
+    });
+
+    String home_path;
+    const char * home_path_cstr = getenv("HOME"); // NOLINT(concurrency-mt-unsafe)
+    if (home_path_cstr)
+        home_path = home_path_cstr;
+
+    if (config().has("history-file"))
+        history_file = config().getString("history-file");
+    else
+        history_file = home_path + "/.keeper-client-history";
+
+    if (!history_file.empty() && !fs::exists(history_file))
+    {
+        try
+        {
+            FS::createFile(history_file);
+        }
+        catch (const ErrnoException & e)
+        {
+            if (e.getErrno() != EEXIST)
+                throw;
+        }
+    }
+
+    Poco::Logger::root().setLevel(config().getString("log-level", "error"));
+
+    EventNotifier::init();
+}
+
+void KeeperClient::executeQuery(const String & query)
+{
+    std::vector<String> queries;
+    boost::algorithm::split(queries, query, boost::is_any_of(";"));
+
+    for (const auto & query_text : queries)
+    {
+        if (!query_text.empty())
+            processQueryText(query_text);
+    }
+}
+
+bool KeeperClient::processQueryText(const String & text)
+{
+    if (exit_strings.find(text) != exit_strings.end())
+        return false;
+
+    std::vector<String> tokens;
+    boost::algorithm::split(tokens, text, boost::is_any_of(" "));
```

Comment:
It does not look correct.
What if the path contains whitespace? Line break? Backslash?
How to quote the values?

Reply:
Yeah, I wrote this before I learned how to use IParser. I suppose we can use it to parse Keeper's "queries" too.

Reply:
Do you think it looks better now?